File: agents/csharp/Core/Executor.cs
using System.Diagnostics;

namespace PhantomAgent.Core;

public class Executor
{
    public string Cwd { get; private set; }

    public Executor()
    {
        Cwd = Directory.GetCurrentDirectory();
    }

    public (string stdout, string stderr, int exitCode) Execute(string command, int timeout = 30, bool usePowerShell = false)
    {
        try
        {
            string fileName;
            string arguments;
            if (usePowerShell)
            {
                fileName = "powershell.exe";
                arguments = $"-NoProfile -NonInteractive -Command \"{command.Replace("\"", "\\\"")}\"";
            }
            else
            {
                fileName = "cmd.exe";
                arguments = $"/c {command}";
            }

            var psi = new ProcessStartInfo
            {
                FileName = fileName,
                Arguments = arguments,
                WorkingDirectory = Cwd,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            };

            using var proc = Process.Start(psi);
            if (proc == null)
                return ("", "Failed to start process", -1);

            // Read both streams to avoid deadlocks
            var stdoutTask = proc.StandardOutput.ReadToEndAsync();
            var stderrTask = proc.StandardError.ReadToEndAsync();

            if (!proc.WaitForExit(timeout * 1000))
            {
                try { proc.Kill(entireProcessTree: true); } catch { }
                return ("", "Command timed out", -1);
            }

            return (stdoutTask.GetAwaiter().GetResult(),
                    stderrTask.GetAwaiter().GetResult(),
                    proc.ExitCode);
        }
        catch (Exception ex)
        {
            return ("", ex.Message, -1);
        }
    }

    public (string path, bool ok) Cd(string path)
    {
        string target = Path.IsPathRooted(path) ? path : Path.Combine(Cwd, path);
        target = Path.GetFullPath(target);
        if (Directory.Exists(target))
        {
            Cwd = target;
            return (target, true);
        }
        return ($"Directory not found: {path}", false);
    }
}

File: agents/python/core/executor.py
import os
import subprocess
import logging

log = logging.getLogger("phantom.core.executor")


class Executor:
    """Command execution engine with working directory tracking."""

    def __init__(self):
        self.cwd = os.getcwd()

    def execute(self, cmd: str, timeout: int = 30) -> tuple[str, str, int]:
        """Run a shell command. Returns (stdout, stderr, exit_code)."""
        try:
            proc = subprocess.Popen(
                cmd, shell=True,
                stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                cwd=self.cwd,
                env=os.environ.copy()
            )
            stdout, stderr = proc.communicate(timeout=timeout)
            return stdout.decode(errors="replace"), stderr.decode(errors="replace"), proc.returncode
        except subprocess.TimeoutExpired:
            proc.kill()
            proc.wait()
            return "", "Command timed out", -1
        except Exception as e:
            log.error(f"Execution failed: {e}")
            return "", str(e), -1

    def cd(self, path: str) -> tuple[str, bool]:
        """Change working directory."""
        target = os.path.join(self.cwd, path) if not os.path.isabs(path) else path
        target = os.path.realpath(target)
        if os.path.isdir(target):
            self.cwd = target
            return target, True
        return f"Directory not found: {path}", False
